File: code/ex128.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int num : nums)    set.add(num);
        
        int res = 0;
        for (int n : set) {    // Traverse the set
            if (!set.contains(n - 1)) {    // Find the starting point of a consecutive chain
                int cnt = 0, s = n;
                while (set.contains(s)) {    // Move upward to find the farthest point this chain can reach
                    s++;
                    cnt++;
                }
                res = Math.max(res, cnt);
            }
        }
        
        return res;
    } 
}

File: code/ex128.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> set(nums.begin(), nums.end());
        int MAX = 0;
        for (int num : nums) {
            int len = 0;
            if (set.count(num - 1) == 0) {
                while (set.count(num) > 0) {
                    len++;    num++;
                }
            }
            MAX = max(MAX, len);
        }
        return MAX;
    }
};